Extending use cases (#23)
temporalio/documentation-legacy: docs/use-cases-orchestration.md

---
id: use-cases-orchestration
title: Microservice Orchestration
sidebar_label: Microservice Orchestration
---

## Motivation

> It is common to break a large application into **microservices** structured around the application's distinct business capabilities. Services are owned by unrelated teams with individual technology choices and release processes.

Reviewer: ```suggestion
It is common to break a large application into **microservices** structured around the application's distinct business capabilities. Services are often owned by unrelated teams who have made their own technology choices and release processes.
```

Each microservice typically has its own storage and interacts with other services via a well-defined API. At the same time, business processes (workflows) may need to call multiple microservices to achieve the desired outcome.

The implementation must **guarantee** that all of the calls eventually succeed even in the case of prolonged failures of individual microservices, preserving business rules across multiple independent sub-domains.

> The service interdependency graph can be remarkably complicated. Processes can run asynchronously, in parallel, some tasks need information from other systems, and the next steps depend on the outcome of previous activities.

Reviewer: ```suggestion
Interdependency graphs between services can be remarkably complicated. Processes can run asynchronously or in parallel, some tasks need information from other systems, and the next steps often depend on the outcome of previous activities.
```

## Benefits of Temporal

> Temporal is a perfect fit for microservice orchestration scenarios. It guarantees that workflow code eventually completes, has built-in support for exponential activity **retries**, and simplifies the coding of the **compensation** logic. 

Reviewer: Link to saga here probably.

> You can define retries, rollbacks, or even a human intervention in case of failure. Workflows are defined in general-purpose programming languages that bring the ultimate **flexibility** in contrast with text-based DSL engines.

Reviewer: ```suggestion
You can define retries, rollbacks, or even a human intervention step in the case of failure. Workflows are defined in general-purpose programming languages that bring the ultimate **flexibility** especially when compared to text-based DSL engines.
```

> Temporal provides full **visibility** into each workflow's state, in contrast to an ad-hoc orchestration based on queues where getting a current status of each request is virtually impossible.

Reviewer: ```suggestion
Temporal provides full **visibility** into each workflow's state. Contrast this with ad-hoc orchestration based on queues where getting a current status of each request is virtually impossible.
```

sidebar_label: Microservice Orchestration
---

## Motivation

> It is common to break a large application into **microservices** structured around the distinct business capabilities. Each microservice typically has its storage and interacts with other services via an API. At the same time, business processes may need to call multiple microservices to achieve the desired outcome.

Reviewer: ```suggestion
It is common to break a large application into **microservices** structured around the applications distinct business capabilities. Each microservice typically has its own storage and interacts with other services via a well defined API. At the same time, business processes (workflows) may need to call multiple microservices to achieve the desired outcome.
```

> The implementation must guarantee that all of the calls must eventually succeed even with the occurrence of prolonged failures of individual microservices.

Reviewer: ```suggestion
The implementation must guarantee that all of the calls will eventually succeed even in the case of prolonged failures of individual microservices.
```

> In some cases, instead of trying to complete the process by retrying for a long time, compensation rollback logic should be executed. [Saga Pattern](https://microservices.io/patterns/data/saga.html) is one way to standardize compensation APIs.

Reviewer: ```suggestion
In some cases, instead of trying to complete the process by continuously retrying, compensation rollback logic should be executed. [Saga Pattern](https://microservices.io/patterns/data/saga.html) is one way to standardize compensation APIs.
```

## Benefits

> Temporal is a perfect fit for such scenarios. It guarantees that workflow code eventually completes, has built-in support for unlimited exponential activity retries, and simplifies the coding of the compensation logic. 

Reviewer: ```suggestion
Temporal is a perfect fit for such scenarios. It guarantees that workflow code eventually completes, has built-in support for exponential activity retries, and simplifies the coding of the compensation logic. 
```

> Temporal also gives full visibility into the state of each workflow, in contrast to an ad-hoc orchestration based on queues where getting a current status of each request is virtually impossible.

Reviewer: ```suggestion
Temporal provides full visibility into the state of each workflow, in contrast to an ad-hoc orchestration based on queues where getting a current status of each request is virtually impossible.
```

## Example

The workflow snippet below orchestrates two activity calls: booking a hotel and reserving a flight. If the first activity fails (including all the configured retries), the workflow returns directly.

> However, if the first reservation succeeds, but the second one fails, you may want to cancel the hotel to avoid undesired charges. That's why the error-handling block contains a third activity call that cancels the hotel reservation before completing the workflow.

Reviewer: ```suggestion
However, if the first reservation succeeds, but the second one fails, you may want to cancel the already booked hotel to avoid undesirable charges. This is why the error-handling block contains a third activity call which cancels the hotel reservation before completing the workflow.
```

```go
// Step 1: Book a hotel.
var hotel HotelResult
err := workflow.ExecuteActivity(ctx, activities.BookHotel, tripRequest).Get(ctx, &hotel)
if err != nil {
    // If the hotel reservation failed, fail the workflow.
    return err
}

// Step 2: Book a flight.
var flight FlightResult
err := workflow.ExecuteActivity(ctx, activities.BookFlight, tripRequest).Get(ctx, &flight)
if err != nil {
    // If the flight reservation failed, cancel the hotel.
    var cancelResult HotelCancellationResult
    cancelErr := workflow.ExecuteActivity(ctx, activities.CancelHotel, hotel).Get(ctx, &cancelResult)
    if cancelErr != nil {
        // The hotel cancellation failed... probably, some manual action is needed.
        return cancelErr
    }
    
    return err
}

// Both reservations succeeded.
```

## Next Steps

> Here are two real-world examples of Temporal-based service orchestration scenarios:

Reviewer: ```suggestion
Here are two real-world examples of Temporal-powered service orchestration scenarios:
```

 * [Using Temporal workflows to spin up Kubernetes](https://banzaicloud.com/blog/introduction-to-cadence/) by Banzai Cloud

>  * Improving the User Experience with Uber’s [Customer Obsession Ticket Routing Workflow and Orchestration Engine](https://eng.uber.com/customer-obsession-ticket-routing-workflow-and-orchestration-engine/)

Reviewer 2: Add link to Java Saga Sample: https://github.com/temporalio/temporal-java-samples/tree/master/src/main/java/io/temporal/samples/bookingsaga
